plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts

```diff
@@ -0,0 +1,364 @@
+            logger.error(`${this.getProviderName()} refresh failed`, error);
+          }
+        },
+      });
+    };
+  }
+
+  private async refresh(logger: Logger): Promise<void> {
+    if (!this.connection) {
+      throw new Error(
+        `Gitlab discovery connection not initialized for ${this.getProviderName()}`,
+      );
+    }
+
+    const client = new GitLabClient({
+      config: this.integration.config,
+      logger: logger,
+    });
+
+    const users = paginated<GitLabUser>(options => client.listUsers(options), {
+      page: 1,
+      per_page: 100,
+      active: true,
+    });
+
+    const groups = paginated<GitLabGroup>(
+      options => client.listGroups(options),
+      {
+        page: 1,
+        per_page: 100,
+      },
+    );
+
+    const idMappedGroup: { [groupId: number]: GitLabGroup } = {};
+
+    const res: Result = {
+      scanned: 0,
+      matches: [],
+    };
+
+    const groupRes: GroupResult = {
+      scanned: 0,
+      matches: [],
+    };
+
+    for await (const group of groups) {
+      if (!this.config.groupPattern.test(group.full_path ?? '')) {
+        continue;
+      }
+
+      groupRes.scanned++;
+      groupRes.matches.push(group);
+
+      idMappedGroup[group.id] = group;
+    }
+
+    for await (const user of users) {
+      if (!this.config.userPattern.test(user.email ?? '')) {
+        continue;
+      }
+
+      res.scanned++;
+
+      if (user.active) {
+        continue;
+      }
+
+      const memberships = await client.getUserMemberships(user.id);
+      const userGroups: GitLabGroup[] = [];
+
+      for (const i of memberships) {
+        if (
+          i.source_type === 'Namespace' &&
+          idMappedGroup.hasOwnProperty(i.source_id)
+        ) {
+          userGroups.push(idMappedGroup[i.source_id]);
+        }
+      }
+
+      user.groups = userGroups;
+
+      res.matches.push(user);
+    }
+
+    const groupsWithUsers = groupRes.matches.filter(group => {
+      return (
+        res.matches.filter(x => {
+          return !!x.groups?.find(y => y.id === group.id);
+        }).length > 0
+      );
+    });
+
+    const userEntities = res.matches.map(p =>
+      this.createUserEntity(p, this.integration.config.host),
+    );
+    const groupEntities = this.createGroupEntities(
+      groupsWithUsers,
+      this.integration.config.host,
+    );
+
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: [...userEntities, ...groupEntities].map(entity => ({
+        locationKey: this.getProviderName(),
+        entity: this.withLocations(this.integration.config.host, entity),
+      })),
+    });
+  }
+
+  private createGroupEntities(
+    groupResult: GitLabGroup[],
+    host: string,
+  ): GroupEntity[] {
+    const idMapped: { [groupId: number]: GitLabGroup } = {};
+    const entities: GroupEntity[] = [];
+
+    for (const group of groupResult) {
+      idMapped[group.id] = group;
+    }
+
+    for (const group of groupResult) {
```

Comment:
this could probably be done in a single loop, unless I'm missing something

Reply:
No, cause of the parent mapping. First loop creates the indexed array, second loop adds the parent relation to the entity.

Reply:
Sorry, I love a challenge. Shouldn't this work if we use lodash:

```
const idMapped = _.groupBy(groupResult, 'id');
const entities = _.map(groupResult, group => {
    const entity = this.createGroupEntity(group, host);
    if (group.parent_id && idMapped[group.parent_id]) {
        entity.spec.parent = idMapped[group.parent_id][0].full_path;
    }
    return entity;
});
```

Feel free to ignore this and resolve again. Just a suggestion! Good work

Reply:
Probably :D, I am not really familiar with lodash. Doesn't it basically do the same? Just a different syntax? I can change it if really necessary ;)